### PR TITLE
Added custom dashboard mechanism

### PR DESCRIPTION
**Pull request by ar4s**
dashboard-create-graph-dc
dashboard-create-dasboard
dashboard-link

### CODE REVIEW

**Comment by ar4s** [comment]
- [x] add title to graph
- [x] group graphs in dashboard 
- [x] add tooltip to each bar

Nice to have:
- [ ] layout based on flex

File: src/ralph/dashboard/models.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+        'horizontalBars': True, 'axisY': {
+            'offset': 70,
+        }
+    })
+    line = _('Line').extra(func='Line')
+
+
+class Graph(models.Model):
+    # TODO: rename to model_path
+    model = models.CharField(max_length=100)
```
Comment (szok) [inline src/ralph/dashboard/models.py]: ContentType will be better

**Comment by mkurek** [comment]
@ar4s could you provide sample json params for above charts?

**Comment by ar4s** [comment]
For first (Back office):
{"labels": "model__name", "series": "model__manufacturer__name", "filters": {"status": 10}}
For second (Data center):
{"labels": "name", "series": "serverroom__rack__datacenterasset", "filters": {}}

**Comment by ghost** [comment]
`ChartType` is strictly related to chartist lib so maybe this could be improved somehow?

**Comment by ar4s** [comment]
https://github.com/allegro/ralph/pull/2202/files#diff-391c867666f52e6647528dd3e9b93c69

File: src/ralph/dashboards/filter_parser.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+    def __init__(self, queryset, filters):
+        self.queryset = queryset
+        self.filters = filters
+
+    def get_queryset(self):
+        parsed_filters = {}
+        for key, value in self.filters.items():
+            params = key.split('|')
+            key = self.transform_key_lookup(params[0])
```
Comment (szok) [inline src/ralph/dashboards/filter_parser.py]: Key should not be under len(params) ==1 ?

File: src/ralph/dashboards/filter_parser.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+    def filter_from_now(self, value):
+        period_mapper = {
+            'd': 'days',
+            'm': 'months',
+            'y': 'years',
+        }
+        val, period = FILTER_FROM_NOW.match(value).groups()
+        result = datetime.date.today() + relativedelta(**{
+            period_mapper.get(period, period): int(val)
```
Comment (szok) [inline src/ralph/dashboards/filter_parser.py]: delete period, period

**Comment by szok** [comment]
:ok:

File: src/ralph/dashboards/models.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+    def get_data(self):
+        model = self.model.model_class()
+        model_manager = model._default_manager
+        aggregate_type = AggregateType.from_id(self.aggregate_type)
+        queryset = model_manager.all()
+        filters = self.params.get('filters', None)
+        if filters:
+            queryset = FilterParser(queryset, filters).get_queryset()
+        aggregate_func = getattr(aggregate_type, 'aggregate_func')
```
Comment (mkurek) [inline src/ralph/dashboards/models.py]: why not `aggregate_func = aggregate_type.aggregate_func` ?

File: src/ralph/dashboards/models.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+        queryset = queryset.values(
+            self.params['labels']
```
Comment (mkurek) [inline src/ralph/dashboards/models.py:52]: labels suggest that this it's plural, but as far as i see it's single label

File: src/ralph/dashboards/renderers.py [hunk 6]
```diff
@@ -0,0 +1,83 @@
+import json
+
+from django.template.loader import render_to_string
+from django.utils.safestring import mark_safe
+
+
+class GraphRenderer(object):
```
Comment (mkurek) [inline src/ralph/dashboards/renderers.py]: add some comment/docstring that it's cooperating with `chartist` library

File: src/ralph/dashboards/templates/dashboard/dashboard.html [hunk 7]
```diff
@@ -0,0 +1,22 @@
+{% load graph_tags staticfiles %}
+
+<html>
+  <head>
+    <script type="text/javascript" src="{% static 'vendor/js/chartist.js' %}"></script>
+    <script type="text/javascript" src="{% static 'js/chartist-plugin-barlabels.js' %}"></script>
+    <link rel="stylesheet" type="text/css" href="{% static "css/ralph.css" %}" />
+    <title>{{ name }}</title>
+    <meta http-equiv="refresh" content="60">
```
Comment (mkurek) [inline src/ralph/dashboards/templates/dashboard/dashboard.html]: shouldn't it be configurable in dashboard?

**Comment by mkurek** [comment]
:ok:

**Pull request merged**